otherlibs/dynlink/byte/dynlink.ml

```diff
@@ -95,45 +95,54 @@ module Bytecode = struct
-      Symtable.check_global_initialized compunit.cu_reloc;
-      Symtable.update_global_table ()
-    with Symtable.Error error ->
-      let new_error : DT.linking_error =
-        match error with
-        | Symtable.Undefined_global s -> Undefined_global s
-        | Symtable.Unavailable_primitive s -> Unavailable_primitive s
-        | Symtable.Uninitialized_global s -> Uninitialized_global s
-        | Symtable.Wrong_vm _ -> assert false
-      in
-      raise (DT.Error (Linking_error (file_name, new_error)))
-    end;
-    (* PR#5215: identify this code fragment by
-       digest of file contents + unit name.
-       Unit name is needed for .cma files, which produce several code
-       fragments. *)
-    let digest = Digest.string (file_digest ^ compunit.cu_name) in
-    let events =
-      if compunit.cu_debug = 0 then [| |]
-      else begin
-        seek_in ic compunit.cu_debug;
-        [| input_value ic |]
-      end in
-    if priv then Symtable.hide_additions old_state;
-    let _, clos = Meta.reify_bytecode code events (Some digest) in
-    try ignore ((clos ()) : Obj.t)
+    let clos = with_lock lock (fun () ->
+        let old_state = Symtable.current_state () in
+        let compunit : Cmo_format.compilation_unit = unit_header in
+        seek_in ic compunit.cu_pos;
+        let code_size = compunit.cu_codesize + 8 in
+        let code = LongString.create code_size in
+        LongString.input_bytes_into code ic compunit.cu_codesize;
+        LongString.set code compunit.cu_codesize (Char.chr Opcodes.opRETURN);
+        LongString.blit_string "\000\000\000\001\000\000\000" 0
+          code (compunit.cu_codesize + 1) 7;
+        begin try
+          Symtable.patch_object code compunit.cu_reloc;
+          Symtable.check_global_initialized compunit.cu_reloc;
+          Symtable.update_global_table ()
+        with Symtable.Error error ->
+          let new_error : DT.linking_error =
+            match error with
+            | Symtable.Undefined_global s -> Undefined_global s
+            | Symtable.Unavailable_primitive s -> Unavailable_primitive s
+            | Symtable.Uninitialized_global s -> Uninitialized_global s
+            | Symtable.Wrong_vm _ -> assert false
+          in
+          raise (DT.Error (Linking_error (file_name, new_error)))
+        end;
+        (* PR#5215: identify this code fragment by
+           digest of file contents + unit name.
+           Unit name is needed for .cma files, which produce several code
+           fragments. *)
+        let digest = Digest.string (file_digest ^ compunit.cu_name) in
+        let events =
+          if compunit.cu_debug = 0 then [| |]
+          else begin
+            seek_in ic compunit.cu_debug;
+            [| input_value ic |]
+          end in
+        if priv then Symtable.hide_additions old_state;
+        let _, clos = Meta.reify_bytecode code events (Some digest) in
+        clos
+      )
+    in
+    try ignore ((clos ()) : Obj.t);
```

Comment:
I think it could be helpful to point out explicitly that this must not run under the lock, because the closure could itself try to use Dynlink. (There is a comment to this effect on the call to `P.run`, but having one here would also help.)

Reply:
Comment(s) added.